fbcore/src/main/java/com/facebook/datasource/RetainingDataSourceSupplier.java

```diff
@@ -0,0 +1,140 @@
+package com.facebook.datasource;
+
+import com.facebook.common.executors.CallerThreadExecutor;
+import com.facebook.common.internal.Supplier;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.concurrent.ThreadSafe;
+
+@NotThreadSafe
+public class RetainingDataSourceSupplier<T> implements Supplier<DataSource<T>> {
+
+  private final Set<RetainingDataSource> mDataSources =
+      Collections.newSetFromMap(new WeakHashMap<RetainingDataSource, Boolean>());
+
+  private Supplier<DataSource<T>> mCurrentDataSourceSupplier = null;
+
+  @Override
+  public DataSource<T> get() {
+    RetainingDataSource dataSource = new RetainingDataSource();
+    dataSource.setSupplier(mCurrentDataSourceSupplier);
+    mDataSources.add(dataSource);
+    return dataSource;
+  }
+
+  public void setSupplier(Supplier<DataSource<T>> supplier) {
+    mCurrentDataSourceSupplier = supplier;
+    for (RetainingDataSource dataSource: mDataSources) {
+      dataSource.setSupplier(supplier);
+    }
+  }
+
+  @ThreadSafe
+  private class RetainingDataSource extends AbstractDataSource<T> {
+    @GuardedBy("RetainingDataSource.this")
+    @Nullable
+    private DataSource<T> mDataSource = null;
+
+    public void setSupplier(@Nullable Supplier<DataSource<T>> supplier) {
+      // early return without calling {@code supplier.get()} in case we are closed
+      if (isClosed()) {
+        return;
+      }
+      DataSource<T> oldDataSource;
+      DataSource<T> newDataSource = (supplier != null) ? supplier.get() : null;
+      synchronized (RetainingDataSource.this) {
+        if (isClosed()) {
```

Comment:
Can we rewrite this s.t. some of the side-effects (e.g. `oldDataSource = newDataSource` will make `newDataSource` being closed below) clearer? For example with an early return:

```java
if (isClosed()) {
  closeSafely(newDataSource);
  return;
}